_posts/pl/2021-12-XX-transactions.md

```diff
@@ -0,0 +1,217 @@
+- **I** (_ang. isolation_) – izolacja – właściwość, która odseparowuje od siebie operacje wykonywane jednocześnie tak, aby te nie miały na siebie wpływu. Chodzi np. o sytuację, gdy odwołujemy się do tej samej tabeli w bazie. Można też powiedzieć, że izolacja zapewnia, że w przypadku wykonywania operacji w sposób równoległy baza danych pozostanie w takim stanie, jakby operacje te zostały wykonane sekwencyjnie.
+
+- **D** (_ang. durability_) – trwałość – gwarancja, że zaakceptowana transakcja nie zniknie niespodziewanie z systemu np. w przypadku awarii. Trwałość uzyskuje się poprzez zastosowanie mechanizmu replikacji, odnotowywania operacji w dzienniku logów czy poprzez zapis danych na dysku.
+
+# Zarządzanie transakcjami w JDBC
+Każdy, kto korzysta z mechanizmu transakcji w Springu wie, że pomocna w tym jest adnotacja **@Transactional**. Dzięki niej możemy zaoszczędzić trochę czasu, bo sami nie musimy implementować całego mechanizmu, którym zarządza framework. Warto jednak wiedzieć co dzieje się „pod maską”. Zobaczmy zatem jak wygląda uproszczone zarządzanie transakcją w JDBC, na którym zawsze bazujemy.
+
+```java
+// nawiązanie połączenia z bazą danych
+Connection connection = dataSource.getConnection();
+
+try (connection) {
+	// Po utworzenia połączenia z bazą, połaczenie to jest w trybie tzw. „automatycznego zatwierdzania”. Oznacza to, że każda pojedyncza instrukcja SQL jest traktowana jako jedna transakcja i jest automatycznie zatwierdzana zaraz po jej wykonaniu. Ustawienie flagi autoCommit na false pozwala na wykonanie więcej niż jednej instrukcji w ramach jednej transakcji.
+	connection.setAutoCommit(false);
+
+	// ...
+	// operacje wchodzące w skład transakcji
+	// ...
+
+	// zatwierdzenie transakcji
+	connection.commit(); 
+} catch (SQLException e) {
+	// wycofanie zmian w przypadku błędów 
+	connection.rollback(); 
+}
+```
+
+## Poziomy izolacji
+Mając już ogólny pogląd tego, jak zaimplementowany jest mechanizm transakcji warto wspomnieć o sposobie ustawiania blokad w dostępie do danych. Sposób ustawiania blokad określany jest mianem **poziomu izolacji** transakcji i pomaga w zachowaniu integralności danych. Określają one zasady równoległej realizacji kilku transakcji.
+
+W JDBC występuje 5 poziomów izolacji transakcji:
+- `TRANSACTION_NONE` - brak transakcji
+  
+- `TRANSACTION_READ_COMMITTED` - inne transakcje nie mogą odczytywać zmienionych wierszy przed wywołaniem metody _commit()_
+  
+- `TRANSACTION_READ_UNCOMMITTED` - dopuszcza odczyt danych przed wywołaniem metody _commit()_
+  
+- `TRANSACTION_REPEATABLE_READ` - chroni przed sytuacją, gdy transakcja odczytuje wiersz, druga transakcja go zmienia, a pierwsza ponownie odczytuje otrzymując inne dane
+  
+- `TRANSACTION_SERIALIZABLE` - chroni przed sytuacją, gdy jedna transakcja odczytuje zbiór wierszy spełniający określone kryteria, a następnie druga transakcja wstawia wiersz spełniający ten sam warunek, po czym pierwsza ponownie odczytuje ten sam zbiór wierszy dostając nowy rekord
```

Comment:
Strasznie to na okrętkę jest wytłumaczone.

Reply:
wywaliłem te opisy, zostawiłem tabelkę i trochę to inaczej opisałem